runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/state/WindmillStateInternalsTest.java

```diff
@@ -763,6 +786,28 @@ public void testMultimapPutAndGet() {
     assertThat(result.read(), Matchers.containsInAnyOrder(1, 1, 2, 3));
   }
 
+  @Test
+  public void testMapViaMultimapPutAndGet() {
```

Comment:
Some other edge cases that might be worth testing (unless I missed them somewhere):

- Overwriting a key with a new value and then reading returns the new value.
- Reading a key that hasn't been "put" returns correct value.
- Put then Remove then Read returns empty.

Reply:
added more tests and it exposed bug in multimapstate (improved multimap state test to expose it directly). During sequence of remove/put/remove/put multimapstate would remove the local knowledge of the remove.